crates/primitives/utils/src/service.rs

```diff
@@ -1,44 +1,441 @@
+//! > service. If you do not, or your service takes longer than
+//! > [SERVICE_GRACE_PERIOD] to shutdown, then your service will be forcefully
+//! > canceled.
+//!
+//! ## Cancellation requests
+//!
+//! Any service with access to a [ServiceContext] can request the cancellation
+//! of _any other service, at any point during execution_. This can be used for
+//! error handling for example, by having a single service shut itself down
+//! without affecting other services, or for administrative and testing purposes
+//! by having a node operator toggle services on and off from a remote endpoint.
+//!
+//! You can use the following methods to request for the cancellation of a
+//! service:
+//!
+//! - [ServiceContext::cancel_global]: cancels all services.
+//! - [ServiceContext::cancel_local]: cancels this service and all its children.
+//! - [ServiceContext::service_remove]: cancel a specific service.
+//!
+//! ## Start requests
+//!
+//! You can _request_ for a service to be restarted by calling
+//! [ServiceContext::service_add]. This is not guaranteed to work, and will fail
+//! if the service is already running or if it has not been registered to
+//! [the set of global services](#service-orchestration) at the start of the
+//! program.
+//!
+//! ## Atomic status checks
+//!
+//! All service updates and checks are performed atomically with the use of
+//! [tokio_util::sync::CancellationToken] and [MadaraServiceMask], which is a
+//! [std::sync::atomic::AtomicU8] bitmask with strong [std::sync::atomic::Ordering::SeqCst]
+//! cross-thread ordering of operations. Services are represented as unique
+//! powers of 2 with [MadaraServiceId]. You can use this to construct your own
+//! mask to check the status of multiple related services at once using
+//! [MadaraServiceMask::status].
+//!
+//! > **Note**
+//! > The use of [std::sync::atomic::AtomicU8] limits the number of possible
+//! > services to 8. This might be increased in the future if new services are
+//! > added.
+//!
+//! ---
+//!
+//! # Service orchestration
+//!
+//! Services are orchestrated by a [ServiceMonitor], which is responsible for
+//! registering services, marking them as active or inactive as well as starting
+//! and restarting them upon request. [ServiceMonitor] also handles the
+//! cancellation of all services upon receiving a `SIGINT` or `SIGTERM`.
+//!
+//! > **Important**
+//! > Services cannot be started or restarted if they have not been registered
+//! > with [ServiceMonitor::with].
+//!
+//! Services are run to completion until no service remains, at which point the
+//! node will automatically shutdown.
+//!
+//! [microservices]: https://en.wikipedia.org/wiki/Microservices
 
 use anyhow::Context;
-use std::{fmt::Display, panic, sync::Arc};
+use futures::Future;
+use serde::{Deserialize, Serialize};
+use std::{
+    fmt::{Debug, Display},
+    panic,
+    sync::Arc,
+    time::Duration,
+};
 use tokio::task::JoinSet;
 
+pub const SERVICE_COUNT: usize = 8;
```

Comment:
i think this shouldnt be here btw, it should be in main.rs

if an appchain wants to add a service by importing all of our crates and just have their own main.rs which starts and hook services differently (/ using other services), this wont work for them

Reply:
Well they'll have bigger problems due to the limit in size of using an `std::sync::atomic::AtomicU8`. I assume you mean having the service id behind a trait so appchains can create their own services?

Reply:
Well, this was a bit of a refactor but `MadaraServiceId` is now only a blanket implementation of `Serviceid` for our services, with `PowerOfTwo` being the actual backing type used to identify services. Appchains can create and register their own services by implementing the `Service` and `ServiceId` traits on their own structs, and they are free to use any service id enum they want, all relevant methods now only require an `impl ServiceId`.

As for `SERVICE_COUNT`, it has been renamed to `SERVICE_COUNT_MAX` to better reflect what it actually represents and has been increased to 64.